### ISSUE TITLE
aws-cloudwatch scaler only works with expressions that return a value greater than `activationTargetMetricValue`

### ISSUE THREAD

**Issue by karl-sprig** [kedacore/keda#3614]
### Proposal

I would like greater control over the logic in the `IsActive()` function for the `aws-cloudwatch` scaler.

For my needs (outlined below) there are a few different ways this could be implemented:

- just a simple way to indicate that the `IsActive()` logic should be negated. This could be a simple `targetMetricComparisonOperator` field which defaults to the current `>` but would also support `<`:

  ```yaml
  # See: https://keda.sh/docs/2.8/scalers/aws-cloudwatch/
  - type: aws-cloudwatch
    <...>
    # We want keda to do $action when the value returned by CW api is < the target
    targetMetricComparisonOperator: "<"
  ```

- support for multiple expressions submitted to the cloudwatch API + a way to indicate which query should be used in the `IsActive()` calculation:

  ```yaml
  triggers:
    - type: aws-cloudwatch
      metadata:
        expressions:
          # Cloudwatch does not support nested queries so we must assign each query an ID which CAN be referenced in other queries
          # First, get the total number of invocations
          # Then compare the results from m1 to the integer 1. If m1 >= 1, return 1. Else, 0.
          - id: m1
            expression: >-
              SELECT COUNT(Invocations) FROM SCHEMA("AWS/Lambda", FunctionName) WHERE FunctionName = 'sacred-cow'
          - id: e1
            expression: >-
              IF(m1 >=, 1, 1, 0)
          # Tell keda to watch only the value of the query identified as `e1` when making decisions about scaling
          targetMetricExpressionID: e1
  ````

- Implement something else entirely using a combination of [Issue #2440](https://github.com/kedacore/keda/issues/2440) and [Issue #3330](https://github.com/kedacore/keda/issues/3330)
  - While both issues are worth perusing/implementing, they don't solve the more local problem: "the cloudwatch scaler only supports a single basic query expression"

**TL;DR:** Cloudwatch explicitly supports multiple queries, the keda scaler should too.

### Use-Case

I have a sacred cow / legacy application that - from time to time - stop running properly.
For reasons that are not important here, this application is off limits.

Fortunately, I have some cloud watch metrics that allow me to see the number times this legacy application has been run per unit time. If the number drops below 1, I need to fire off a simple reboot script which - while blunt - does work reliably.

I have tried to implement this basic "if the count drops to less than 1, run the action" pattern with this:

```yaml
apiVersion: keda.sh/v1alpha1
kind: ScaledJob
  <...>
spec:
  # Check every 60s
  pollingInterval: 60
  # At no point do we need more than 1 power-cycler instance running
  minReplicaCount: 0
  maxReplicaCount: 1

  triggers:
    # See: https://keda.sh/docs/2.8/scalers/aws-cloudwatch/
    - type: aws-cloudwatch
      metadata:
        namespace: AWS/Lambda
        metricName: Invocations
        expression: >-
          SELECT MIN(Invocations) FROM SCHEMA("AWS/Lambda", FunctionName, Resource) WHERE FunctionName = 'sacred-cow' AND Resource = 'sacred-cow'

        metricStatPeriod: "60"
        metricUnit: "Count"
        metricCollectionTime: "300"
        minMetricValue: "0"
        targetMetricValue: "1"
        activationTargetMetricValue: "0"

  jobTargetRef:
    <...>
```

However, after some conversation with @JorTurFer on [slack](https://kubernetes.slack.com/archives/CKZJ36A5D/p1661880282181369), this may not be supported directly.

In short, the `expression` will always return `1` when things are nominal. Only when there is no data or the number of invocations per unit time is 0 will the `expression` return 0.

A quick 'hack' was suggested: Invert the logic; return `0` when things are nominal and `1` when things are not.

This is only possible when using [logical operators in your cloudwatch query](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/using-metric-math.html#using-IF-expressions). **However** this requires submititng _multiple_ queries to cloudwatch... which the current `aws-cloudwatch` scaler does not support.

To achieve the 'negate the expression result' workaround from above, you must submit an API query with a payload similar to:

```json
{
    "metrics": [
        [ { "expression": "SELECT COUNT(Invocations) FROM SCHEMA(\"AWS/Lambda\", FunctionName) WHERE FunctionName = 'sacred-cow'", "label": "Metric1", "id": "mInvocationCount", "period": 60, "stat": "SampleCount" } ],
        [ { "expression": "IF(mInvocationCount > 0, 0, 1)", "label": "Expression1", "id": "e1" } ]
    ],
    "view": "timeSeries",
    "stacked": false,
    "region": "us-east-1",
    "stat": "SampleCount",
    "period": 60
}

```

However, the current implementation does not permit this; there is only ever one query expression sent to AWS:

https://github.com/kedacore/keda/blob/96e5bd25b183e950ed87660ee5ebc0a85dca873a/pkg/scalers/aws_cloudwatch_scaler.go#L349

**Comment by JorTurFer** [comment]
I answered you but let's talk about this here to track the conversation.
why do you need 2 queries? why not just `if(select count(invocations)>0, 1,0)` ?
I have seen that selecting the symbol used for the activation could be enough to you. AFAIK AWS CloudWatch supports arithmetic operations, just multiply you value for `-1` and you have already changed the symbol in the upstream, any change is needed in KEDA side. Could it work?

**Comment by karl-sprig** [comment]
> I'm not an expert in AWS and maybe this is not doable...

Not to worry. AWS does not explicitly spell this out anywhere that I can find. It's implied from _all_ of [the examples](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/using-metric-math.html) though.  You can also confirm this experimentally as i'll show below.

**Cloudwatch does not support nested queries in the form of `Select SUM($thing) where $thing_id in (Select from ....)`**

> AFAIK AWS CloudWatch supports arithmetic operations, just multiply you value for -1 and you have already changed the symbol in the upstream, any change is needed in KEDA side. Could it work?

Same with math operations. You get a _single_ expression per query `id` and then you can compose the `id`s with operators but you can not use an operator _and_ a 'normal' query in the same expression:

![image](https://user-images.githubusercontent.com/103214175/187734682-50ac1ec6-632a-469e-915f-0aee496d752e.png)


**versus**

![image](https://user-images.githubusercontent.com/103214175/187735856-b7a3dc27-dd02-43ce-91f8-f144433d63a9.png)



I hope that clarifies things

**Comment by JorTurFer** [comment]
mmm, in that case, seems that ATM your request is not supported by KEDA, sorry.
Once formulas feature is ready this could work, but I don't have any idea for the moment :(

**Comment by JorTurFer** [comment]
one thing that you could do is create your own RestAPI/gRCP service which queries cloudwatch and applies the formulas that you need, and use Metrics API Scaler or External Scaler to query your service from KEDA
